pkg/routeagent_driver/handlers/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,331 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthchecker
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/submariner-io/admiral/pkg/log"
+	"github.com/submariner-io/admiral/pkg/resource"
+	"github.com/submariner-io/admiral/pkg/util"
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	v1typed "github.com/submariner-io/submariner/pkg/client/clientset/versioned/typed/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/event"
+	"github.com/submariner-io/submariner/pkg/pinger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var RouteAgentUpdateInterval = 60 * time.Second
+
+const UpdateTimestampAnnotation = "update-timestamp"
+
+type Config struct {
+	PingInterval         uint
+	MaxPacketLossCount   uint
+	HealthCheckerEnabled bool
+	NewPinger            func(pinger.Config) pinger.Interface
+}
+
+type controller struct {
+	event.HandlerBase
+	sync.RWMutex
+	pingers        map[string]pinger.Interface
+	localNodeName  string
+	remoteEndpoint map[string]*submarinerv1.Endpoint
+	version        string
+	config         *Config
+	stopCh         chan struct{}
+	client         v1typed.RouteAgentInterface
+}
+
+var logger = log.Logger{Logger: logf.Log.WithName("HealthChecker")}
+
+func NewHealthCheckerHandler(config *Config, client v1typed.RouteAgentInterface, version, nodeName string) event.Handler {
+	controller := &controller{
+		pingers:        map[string]pinger.Interface{},
+		remoteEndpoint: map[string]*submarinerv1.Endpoint{},
+		config:         config,
+		version:        version,
+		client:         client,
+		stopCh:         make(chan struct{}),
+		localNodeName:  nodeName,
+	}
+
+	return controller
+}
+
+func (h *controller) Stop() error {
+	h.Lock()
+	defer h.Unlock()
+
+	for _, p := range h.pingers {
+		p.Stop()
+	}
+
+	h.pingers = map[string]pinger.Interface{}
+
+	close(h.stopCh)
+
+	err := h.client.Delete(context.TODO(),
+		h.localNodeName, metav1.DeleteOptions{})
+	if err != nil {
+		logger.Warningf("Error deleting RouteAgent: %s: %v", h.localNodeName, err)
+	}
+
+	return nil
+}
+
+func (h *controller) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	if !h.config.HealthCheckerEnabled && h.State().IsOnGateway() {
```

Comment:
Shouldn't this be OR?

```suggestion
	if !h.config.HealthCheckerEnabled || h.State().IsOnGateway() {
```

Reply:
Done